app/test/process_proposal_test.go

```diff
@@ -164,13 +165,76 @@ func TestMessageInclusionCheck(t *testing.T) {
 	}
 }
 
+func TestProcessMessagesWithReservedNamespaces(t *testing.T) {
+	testApp := testutil.SetupTestAppWithGenesisValSet(t)
+	encConf := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+
+	signer := testutil.GenerateKeyringSigner(t, testAccName)
+
+	type test struct {
+		name           string
+		namespace      namespace.ID
+		expectedResult abci.ResponseProcessProposal_Result
+	}
+
+	tests := []test{
+		{"transaction namespace id for message", consts.TxNamespaceID, abci.ResponseProcessProposal_REJECT},
+		{"evidence namespace id for message", consts.EvidenceNamespaceID, abci.ResponseProcessProposal_REJECT},
+		{"tail padding namespace id for message", consts.TailPaddingNamespaceID, abci.ResponseProcessProposal_REJECT},
+		//{"parity shares namespace id for message", consts.ParitySharesNamespaceID, abci.ResponseProcessProposal_REJECT},
```

Comment:
Issue:

Reply:
can we still check for that without actually causing the panic? I _think_ its possible is the message is placed perfectly as the last share in the original square.

Reply:
yes, it is actually possible. But  would make the test a bit ugly I guess.
Currently, the panic happens when calculating the data availability root for the input block. If we skip that step, it will catch the wrong namespace id. 
To do that, we can add a parameter to the test struct called `calculateDARoot` of type `bool`. 
Something like this:

```golang
	type test struct {
		name            string
		namespace       namespace.ID
		calculateDARoot bool
		expectedResult  abci.ResponseProcessProposal_Result
	}

	tests := []test{
		{"transaction namespace id for message", consts.TxNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"evidence namespace id for message", consts.EvidenceNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"tail padding namespace id for message", consts.TailPaddingNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"parity shares namespace id for message", consts.ParitySharesNamespaceID, false, abci.ResponseProcessProposal_REJECT},
		// this case is commented because the shares are ordered. Then, when adding a message with namespace consts.ParitySharesNamespaceID,
		// it panics as it is higher than the maximum allowed namespace ID.
		// issue:
		{"namespace id 200 for message", namespace.ID{0, 0, 0, 0, 0, 0, 0, 200}, true, abci.ResponseProcessProposal_REJECT},
		{"correct namespace id for message", namespace.ID{3, 3, 2, 2, 2, 1, 1, 1}, true, abci.ResponseProcessProposal_ACCEPT},
	}

	for _, tt := range tests {
		pfd, msg := genRandMsgPayForDataForNamespace(t, signer, 8, tt.namespace)
		input := abci.RequestProcessProposal{
			BlockData: &core.Data{
				Txs: [][]byte{
					buildTx(t, signer, encConf.TxConfig, pfd),
				},
				Messages: core.Messages{
					MessagesList: []*core.Message{
						{
							NamespaceId: pfd.GetMessageNamespaceId(),
							Data:        msg,
						},
					},
				},
				OriginalSquareSize: 8,
			},
		}
		if tt.calculateDARoot { <= here
			data, err := coretypes.DataFromProto(input.BlockData)
			require.NoError(t, err)

			shares, _, err := data.ComputeShares(input.BlockData.OriginalSquareSize)
			require.NoError(t, err)

			rawShares := shares.RawShares()

			require.NoError(t, err)
			eds, err := da.ExtendShares(input.BlockData.OriginalSquareSize, rawShares)
			require.NoError(t, err)
			dah := da.NewDataAvailabilityHeader(eds)
			input.Header.DataHash = dah.Hash()
		}
		res := testApp.ProcessProposal(input)
		assert.Equal(t, tt.expectedResult, res.Result)
	}
```

Reply:
if you like it this way, I can update the test

Reply:
nah, as you mentioned in https://github.com/celestiaorg/celestia-app/issues/596#issuecomment-1205785617, this specific case isn't reachable

Reply:
But now that I think about it, I guess it's worth it. So that if someday, for some reason, we're no longer checking for parity shares namespace, we have this test to know exactly what's the issue?
I am okey with both, whichever you prefer.

Reply:
> But now that I think about it, I guess it's worth it. So that if someday, for some reason, we're no longer checking for parity shares namespace, we have this test to know exactly what's the issue?

that's another great point. before reading your comment in #596, I was going to suggest that we split it off into its own test for this exact reason. Its a super explicit reminder of this edge case for the future generations to remember.

Reply:
then we can leave the original test nice and clleeeaaaannnn

Reply:
sounds awesome. Will do.